lib/ansible/modules/source_control/gitlab_tag.py

```diff
@@ -0,0 +1,268 @@
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.pycompat24 import get_exception
+
+
+class GitLabTag(object):
+    def __init__(self, module, git):
+        self._module = module
+        self._gitlab = git
+        self.projectObject = None
+        self.tagObject = None
+
+    def existsTag(self, name):
+        """Validate is tag exists"""
+        project = self.projectObject
+
+        tags = project.tags.list()
+        if len(tags) >= 1:
+            for tag in tags:
+                if tag.name == name:
+                    self.tagObject = tag
+                    return True
+        return False
+
+    def deleteTag(self):
+        """Delete a tag"""
+        tag = self.tagObject
+        if tag.delete():
+            return True
+        else:
+            return False
+
+    def createTag(self, name, ref, description):
+        """Create a tag"""
+        project = self.projectObject
+        tag = self.tagObject
+
+        if tag is None:
+            create_tag = project.tags.create({'tag_name': name, 'ref': ref})
+            create_tag.set_release_description(description)
+            return True
+        return False
+
+    def getProject(self, group, project):
+        """Validates if a project exists."""
+        project_name = project
+        projects = self._gitlab.projects.search(project_name)
+        if len(projects) >= 1:
+            for project in projects:
+                if project.namespace.name == group:
+                    self.projectObject = project
+                    return True
+        return False
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            validate_certs=dict(required=False, default=True, type='bool', aliases=['verify_ssl']),
+            login_user=dict(required=False, no_log=True),
+            login_password=dict(required=False, no_log=True),
+            login_token=dict(required=False, no_log=True),
+            group=dict(required=True),
+            project=dict(required=True),
+            name=dict(required=True),
+            ref=dict(required=False, default="master"),
+            description=dict(required=False),
+            state=dict(default="present", choices=["present", 'absent']),
+        ),
+        supports_check_mode=True
+    )
+
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
+
+    server_url = module.params['server_url']
+    verify_ssl = module.params['validate_certs']
+    login_user = module.params['login_user']
+    login_password = module.params['login_password']
+    login_token = module.params['login_token']
+    project = module.params['project']
+    group = module.params['group']
+    name = module.params['name']
+    ref = module.params['ref']
+    description = module.params['description']
+    state = module.params['state']
+    use_credentials = None
```

Comment:
A lot of this code is duplicated between this `gitlab_tag` module and the `gitlab_hook` module that you've also submitted: https://github.com/ansible/ansible/pull/19093

I think you should extract the common code into a class that goes into `ansible.module_utils`.